ladybug/src/main/java/nl/nn/ibistesttool/MessageEncoder.java

```diff
@@ -0,0 +1,47 @@
+/*
+   Copyright 2021 WeAreFrank!
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package nl.nn.ibistesttool;
+
+import nl.nn.adapterframework.stream.Message;
+import nl.nn.testtool.Checkpoint;
+import nl.nn.testtool.MessageEncoderImpl;
+
+public class MessageEncoder extends MessageEncoderImpl {
+
+	@Override
+	public ToStringResult toString(Object message) {
+		if (message instanceof Message) {
+			// Hide/remove the Message class/object
+			Message m = ((Message)message);
+			ToStringResult toStringResult;
+			if (m.requiresStream()) {
+				toStringResult = new ToStringResult("Waiting for stream to be captured and closed...", null,
+						m.asObject().getClass().getTypeName());
+			} else {
+				toStringResult = super.toString(m.asObject());
+			}
+			return toStringResult;
+		} else {
+			return super.toString(message);
+		}
+	}
+
+	@Override
+	public Object toObject(Checkpoint checkpoint) {
+		return Message.asMessage(super.toObject(checkpoint));
+	}
```

Comment:
Zou je bij deze method het originele bericht mee kunnen geven? Dus in feite het bericht dat vervangen moet worden? Dan kan je aan de hand daarvan bepalen hoe dat het beste kan. Het is dan mogelijk om hetzelfde type er uit te laten komen als er in gaat. Je kan dan de kludge met returnType in debugSenderInputOutputAbort() voorkomen.

Reply:
Wat bedoel je precies? Welke kludge? Het doet van deze methode is om een string representatie (zoals die aan de gebruiker wordt getoond en zoals die wordt opgeslagen in storage implementaties zoals de XmlStorage) weer terug om te zetten naar het oorspronkelijke object. Oorspronkelijk gaf ik String message mee maar omdat je soms net wat meer meta informatie nodig hebt geeft ik nu het Checkponit object mee. Daarop kun je uiteraard een getMessage() doen om de message in String representatie op te halen (degene die ik eerder meegaf) en moet worden omgezet naar het oorspronkelijke object. Zie de default implementatie in Ladybug zelf:

https://github.com/ibissource/ibis-ladybug/blob/master/src/main/java/nl/nn/testtool/MessageEncoderImpl.java#L58

Reply:
Het idee is om het oorspronkelijke object ook mee te geven, zodat je bijvoorbeeld kan zien of dat een Message of een PipeRunResult was.
Het resultaat van deze toObject() wordt gebruikt om een ander object te vervangen. Om dat object gaat het.

Reply:
Done

Reply:
Wanneer wordt nu welke van de twee toObject() methods gebruikt? Ik zou  er maar één verwachten.

Reply:
Dat wordt beschreven in de interface bij het Ladybug project:

https://github.com/ibissource/ibis-ladybug/blob/master/src/main/java/nl/nn/testtool/MessageEncoder.java#L67
https://github.com/ibissource/ibis-ladybug/blob/master/src/main/java/nl/nn/testtool/MessageEncoder.java#L76

Waar het op neerkomt is dat je om een rerun te starten het bericht van het eerste checkpoint van het originele project nodig hebt om met dat bericht de adapter/pipeline opnieuw af te trappen (je hebt dan nog geen message om te stubben).